scarf/merge.py

```diff
@@ -0,0 +1,799 @@
+        self.prepend_text = prepend_text
+        self.reset_cell_filter = reset_cell_filter
+        self.seed = seed
+        self.unique_assays = self.get_unique_assays()
+        self.n_unique_assays = len(self.unique_assays)
+        self.merge_generators = self.create_merge_generators()
+
+    def get_unique_assays(self) -> List[str]:
+        """
+        Get unique assays from both datasets
+        """
+        unique_assays = set()
+        for ds in self.datasets:
+            unique_assays.update(ds.assay_names)
+        return list(unique_assays)
+
+    def create_merge_generators(self) -> List[AssayMerge]:
+        """
+        Create AssayMerge objects for each unique assay
+        """
+        gens = []
+        for assay in self.unique_assays:
+            assay_list = []
+            for ds in self.datasets:
+                if assay in ds.assay_names:
+                    assay_list.append(ds.get_assay(assay))
+                else:
+                    # Generate a dummy assay on the fly
+                    dummy_assay: DummyAssay = self.generate_dummy_assay(ds, assay)
+                    assay_list.append(dummy_assay)
+            gens.append(
+                AssayMerge(
+                    zarr_path=self.zarr_path,
+                    assays=assay_list,
+                    names=self.names,
+                    merge_assay_name=assay,
+                    in_workspaces=self.in_workspaces,
+                    out_workspace=self.out_workspace,
+                    chunk_size=self.chunk_size,
+                    dtype=self.dtype,
+                    overwrite=self.overwrite,
+                    prepend_text=self.prepend_text,
+                    reset_cell_filter=self.reset_cell_filter,
+                    seed=self.seed,
+                )
+            )
+        return gens
+
+    def generate_dummy_assay(self, ds: DataStore, assay_name: str) -> DummyAssay:
+        """
+        Generate a dummy assay for a datastore that doesn't have the specified assay
+        """
+        # Find a datastore that has this assay to get feature information
+        reference_ds = next(
+            ds_ for ds_ in self.datasets if assay_name in ds_.assay_names
+        )
+        reference_assay = reference_ds.get_assay(assay_name)
+
+        reference_chunk = [
+            ds.get_assay(ref).rawData.chunksize[0] for ref in ds.assay_names
+        ]
+        # check if entries in reference_chunk are the same
+        if not all(x == reference_chunk[0] for x in reference_chunk):
+            rowChunkShape = reference_chunk[0]
+        else:
+            rowChunkShape = max(reference_chunk)
+        colChunk = reference_assay.rawData.chunksize[1]
+        chunkShape = (rowChunkShape, colChunk)
+
+        # Create a dummy assay with zero counts and matching features
+        dummy_shape = (ds.cells.N, reference_assay.feats.N)
+        dummy_counts = np.zeros(dummy_shape, dtype=reference_assay.rawData.dtype)
```

Comment:
what is the memory implication of this? Will memory be allocated for this?

Reply:
I have modified this, now we are using zarr.zeros() instead of np.zeros(). Resulting memory footprint is small (347 bytes compared to 1.1G for a (10000, 30000) array of type float32, as an example). 
All operations are in memory. Disk space will be allocated when dumping.